File: 18.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

type pos struct{ x, y, z int }

func (a pos) max(b pos) pos {
	return pos{max(a.x, b.x), max(a.y, b.y), max(a.z, b.z)}
}

func (a pos) min(b pos) pos {
	return pos{min(a.x, b.x), min(a.y, b.y), min(a.z, b.z)}
}

type posset map[pos]bool

func (ps posset) contains(p pos) bool {
	_, found := ps[p]
	return found
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	cubes := posset{}
	for scanner.Scan() {
		var cube pos
		fmt.Sscanf(scanner.Text(), "%d,%d,%d", &cube.x, &cube.y, &cube.z)
		cubes[cube] = true
	}
	surface := 0
	bounds := []pos{pos{}, pos{}}
	for cube, _ := range cubes {
		if !cubes.contains(pos{cube.x + 1, cube.y, cube.z}) {
			surface += 1
		}
		if !cubes.contains(pos{cube.x - 1, cube.y, cube.z}) {
			surface += 1
		}
		if !cubes.contains(pos{cube.x, cube.y - 1, cube.z}) {
			surface += 1
		}
		if !cubes.contains(pos{cube.x, cube.y + 1, cube.z}) {
			surface += 1
		}
		if !cubes.contains(pos{cube.x, cube.y, cube.z - 1}) {
			surface += 1
		}
		if !cubes.contains(pos{cube.x, cube.y, cube.z + 1}) {
			surface += 1
		}
		bounds[0] = bounds[0].min(cube)
		bounds[1] = bounds[1].max(cube)
	}
	bounds[0].x -= 1
	bounds[0].y -= 1
	bounds[0].z -= 1
	bounds[1].x += 1
	bounds[1].y += 1
	bounds[1].z += 1
	external := posset{}
	queue := []pos{bounds[0]}
	for len(queue) > 0 {
		c := queue[0]
		queue = queue[1:]
		if external.contains(c) || cubes.contains(c) ||
			c.x < bounds[0].x || c.y < bounds[0].y || c.z < bounds[0].z ||
			c.x > bounds[1].x || c.y > bounds[1].y || c.z > bounds[1].z {
			continue
		}
		external[c] = true
		queue = append(queue, pos{c.x - 1, c.y, c.z})
		queue = append(queue, pos{c.x + 1, c.y, c.z})
		queue = append(queue, pos{c.x, c.y - 1, c.z})
		queue = append(queue, pos{c.x, c.y + 1, c.z})
		queue = append(queue, pos{c.x, c.y, c.z - 1})
		queue = append(queue, pos{c.x, c.y, c.z + 1})
	}
	esurface := 0
	for c, _ := range cubes {
		if external.contains(pos{c.x + 1, c.y, c.z}) {
			esurface += 1
		}
		if external.contains(pos{c.x - 1, c.y, c.z}) {
			esurface += 1
		}
		if external.contains(pos{c.x, c.y + 1, c.z}) {
			esurface += 1
		}
		if external.contains(pos{c.x, c.y - 1, c.z}) {
			esurface += 1
		}
		if external.contains(pos{c.x, c.y, c.z + 1}) {
			esurface += 1
		}
		if external.contains(pos{c.x, c.y, c.z - 1}) {
			esurface += 1
		}
	}
	fmt.Println(surface, esurface)
}

File: 18.swift
import Foundation

struct Pos: Hashable {
    var x, y, z: Int
}

func max(_ a: Pos, _ b: Pos) -> Pos {
    return Pos(x: max(a.x, b.x), y: max(a.y, b.y), z: max(a.z, b.z))
}

func min(_ a: Pos, _ b: Pos) -> Pos {
    return Pos(x: min(a.x, b.x), y: min(a.y, b.y), z: min(a.z, b.z))
}

var cubes = Set<Pos>()
while let line = readLine() {
    let n = line.numbers
    cubes.insert(Pos(x: n[0], y: n[1], z: n[2]))
}
var surface = 0
var bounds = [Pos(x: 0, y: 0, z: 0), Pos(x: 0, y: 0, z: 0)]
for cube in cubes {
    if !cubes.contains(Pos(x: cube.x + 1, y: cube.y, z: cube.z)) { surface += 1 }
    if !cubes.contains(Pos(x: cube.x - 1, y: cube.y, z: cube.z)) { surface += 1 }
    if !cubes.contains(Pos(x: cube.x, y: cube.y - 1, z: cube.z)) { surface += 1 }
    if !cubes.contains(Pos(x: cube.x, y: cube.y + 1, z: cube.z)) { surface += 1 }
    if !cubes.contains(Pos(x: cube.x, y: cube.y, z: cube.z - 1)) { surface += 1 }
    if !cubes.contains(Pos(x: cube.x, y: cube.y, z: cube.z + 1)) { surface += 1 }
    bounds[0] = min(bounds[0], cube)
    bounds[1] = max(bounds[1], cube)
}
bounds[0].x -= 1; bounds[0].y -= 1; bounds[0].z -= 1
bounds[1].x += 1; bounds[1].y += 1; bounds[1].z += 1

var external = Set<Pos>()
var queue = [bounds[0]]
while !queue.isEmpty {
    let c = queue.removeFirst()
    if external.contains(c) || cubes.contains(c) ||
       c.x < bounds[0].x || c.y < bounds[0].y || c.z < bounds[0].z ||
       c.x > bounds[1].x || c.y > bounds[1].y || c.z > bounds[1].z {
        continue
    }
    external.insert(c)
    queue.append(Pos(x: c.x - 1, y: c.y, z: c.z))
    queue.append(Pos(x: c.x + 1, y: c.y, z: c.z))
    queue.append(Pos(x: c.x, y: c.y - 1, z: c.z))
    queue.append(Pos(x: c.x, y: c.y + 1, z: c.z))
    queue.append(Pos(x: c.x, y: c.y, z: c.z - 1))
    queue.append(Pos(x: c.x, y: c.y, z: c.z + 1))
}
var esurface = 0
for c in cubes {
    if external.contains(Pos(x: c.x + 1, y: c.y, z: c.z)) { esurface += 1 }
    if external.contains(Pos(x: c.x - 1, y: c.y, z: c.z)) { esurface += 1 }
    if external.contains(Pos(x: c.x, y: c.y + 1, z: c.z)) { esurface += 1 }
    if external.contains(Pos(x: c.x, y: c.y - 1, z: c.z)) { esurface += 1 }
    if external.contains(Pos(x: c.x, y: c.y, z: c.z + 1)) { esurface += 1 }
    if external.contains(Pos(x: c.x, y: c.y, z: c.z - 1)) { esurface += 1 }
}
print(surface, esurface)

extension String {
    var numbers: [Int] {
        guard let regex = try? NSRegularExpression(pattern: "-?\\d+", options: [.caseInsensitive]) else {
            return []
        }
        let matches  = regex.matches(in: self, options: [], range: NSMakeRange(0, self.count))
        return matches.map { match in
            String(self[Range(match.range, in: self)!])
        }.map {
            Int($0)!
        }
    }
}
